### PR TITLE
dotCMS/core#16461 Improve tag field and update tests 

### PR DESCRIPTION
**Pull request by fmontes**

### CODE REVIEW

File: projects/dotcms-field-elements/src/utils/test/utils.tsx [hunk 1]
```diff
@@ -10,6 +10,7 @@ export const dotTestUtil = {
         emptyRequiredPristine: ['dot-required', 'dot-invalid', 'dot-untouched', 'dot-pristine'],
         filled: ['dot-valid', 'dot-touched', 'dot-dirty'],
         filledRequired: ['dot-required', 'dot-valid', 'dot-touched', 'dot-dirty'],
-        filledRequiredPristine: ['dot-required', 'dot-valid', 'dot-untouched', 'dot-pristine']
+        filledRequiredPristine: ['dot-required', 'dot-valid', 'dot-untouched', 'dot-pristine'],
+        touchedPristine: ['dot-valid', 'dot-pristine', 'dot-touched']
```
Comment (fmontes) [inline projects/dotcms-field-elements/src/utils/test/utils.tsx]: This one is for when the user focus on an input inside the field and then blur.

File: projects/dotcms-field-elements/src/components/dot-tags/components/dot-autocomplete/dot-autocomplete.e2e.ts [hunk 2]
```diff
@@ -4,197 +4,230 @@ describe('dot-autocomplete', () => {
...
             await page.waitForChanges();
+            const focus = await element.find('*:focus');
 
-            expect(await page.find('li')).toBeNull();
-            expect(spySelectionEvent).toHaveReceivedEventDetail('tag-1');
+            expect(/autoComplete[0-9]{13}/gm.test(focus.getAttribute('id'))).toBe(true);
         });
 
-        it('should trigger lost focus event', async (done) => {
-            element.spyOnEvent('lostFocus').then(() => {
```
Comment (fmontes) [inline projects/dotcms-field-elements/src/components/dot-tags/components/dot-autocomplete/dot-autocomplete.e2e.ts]: This doesn't work if I remove the:

```await input.triggerEvent('blur');```

The test keeps passing.

File: projects/dotcms-field-elements/src/components/dot-tags/components/dot-autocomplete/dot-autocomplete.e2e.ts [hunk 3]
```diff
@@ -4,197 +4,230 @@ describe('dot-autocomplete', () => {
...
 
-                expect(await element.find('input')).toBeDefined();
+                const lis = await element.findAll('ul li');
+                expect(lis.length).toBe(3);
             });
         });
+
+        xdescribe('debounce', () => {});
+
+        xdescribe('data', () => {});
```
Comment (freddyucv) [inline projects/dotcms-field-elements/src/components/dot-tags/components/dot-autocomplete/dot-autocomplete.e2e.ts]: I think we already have this test here

https://github.com/dotCMS/core-web/blob/175c864cebb2bd72595067c141a165387bed6053/projects/dotcms-field-elements/src/components/dot-tags/components/dot-autocomplete/dot-autocomplete.e2e.ts#L74

File: projects/dotcms-field-elements/src/utils/test/utils.tsx (hunk 1 above)
Comment (freddyucv) [reply projects/dotcms-field-elements/src/utils/test/utils.tsx]: where Is it use?

File: projects/dotcms-field-elements/src/components/dot-tags/components/dot-autocomplete/dot-autocomplete.e2e.ts [hunk 4]
```diff
@@ -4,197 +4,230 @@ describe('dot-autocomplete', () => {
...
-            expect(ul.innerHTML).toEqualHtml(`
-                <li data-result="tag-1" class="autoComplete_result" tabindex="1">
-                    <span class="autoComplete_highlighted">t</span>ag-1
-                </li>
-            `);
-        });
+        describe('select', () => {
+            it('should trigger when press enter', async () => {
+                await input.type('test');
+                input.press('Enter');
```
Comment (alfredo-dotcms) [inline projects/dotcms-field-elements/src/components/dot-tags/components/dot-autocomplete/dot-autocomplete.e2e.ts]: add `await`

File: projects/dotcms-field-elements/src/components/dot-tags/components/dot-autocomplete/dot-autocomplete.e2e.ts [hunk 5]
```diff
@@ -4,197 +4,230 @@ describe('dot-autocomplete', () => {
...
-            const ul = await element.find('ul');
-            expect(ul.innerHTML).toEqualHtml('');
-        });
+            it('should trigger when keyboard select a option', async () => {
+                input.type('res');
+                await page.waitForChanges();
 
-        it('should clean options when lost focus', async (done) => {
-            element.spyOnEvent('lostFocus').then(async () => {
+                element.press('ArrowDown');
```
Comment (alfredo-dotcms) [inline projects/dotcms-field-elements/src/components/dot-tags/components/dot-autocomplete/dot-autocomplete.e2e.ts]: add `await`

File: projects/dotcms-field-elements/src/components/dot-tags/components/dot-autocomplete/dot-autocomplete.e2e.ts [hunk 6]
```diff
@@ -4,197 +4,230 @@ describe('dot-autocomplete', () => {
...
-                expect(options.length).toBe(0);
-                done();
-            });
 
-            await input.triggerEvent('blur');
+                expect(spySelectEvent).toHaveReceivedEventDetail('result-3');
+            });
         });
 
+        xdescribe('lostFocus', () => {
```
Comment (alfredo-dotcms) [inline projects/dotcms-field-elements/src/components/dot-tags/components/dot-autocomplete/dot-autocomplete.e2e.ts]: remove?

File: projects/dotcms-field-elements/src/components/dot-tags/components/dot-autocomplete/dot-autocomplete.e2e.ts [hunk 7]
```diff
@@ -4,197 +4,230 @@ describe('dot-autocomplete', () => {
...
-            await input.press('t');
-            await input.press('e');
-            await input.press('s');
-            await input.press('t');
-            await input.press('Enter');
+        it('should clear the result list on esc key', async () => {
+            expect(await input.getProperty('value')).toBe('res');
+            expect(lis.length).toBe(5);
+
+            input.press('Escape');
```
Comment (alfredo-dotcms) [inline projects/dotcms-field-elements/src/components/dot-tags/components/dot-autocomplete/dot-autocomplete.e2e.ts]: add `await`

File: projects/dotcms-field-elements/src/components/dot-tags/components/dot-autocomplete/dot-autocomplete.e2e.ts [hunk 8]
```diff
@@ -4,197 +4,230 @@ describe('dot-autocomplete', () => {
...
         });
 
-        it('should trigger selection event when click a option', async () => {
-            await input.press('t');
-            await page.waitForChanges();
-
-            const option = await page.find('li');
-            await option.press('Enter');
+        it('should focus on the input after item select', async () => {
+            element.press('ArrowDown');
```
Comment (alfredo-dotcms) [inline projects/dotcms-field-elements/src/components/dot-tags/components/dot-autocomplete/dot-autocomplete.e2e.ts]: add `await`

File: projects/dotcms-field-elements/src/components/dot-tags/components/dot-autocomplete/dot-autocomplete.tsx [hunk 9]
```diff
@@ -81,25 +87,27 @@ export class DotAutocompleteComponent {
...
 
-    private getInputElement(): HTMLElement {
+    private getInputElement(): HTMLInputElement {
         return this.el.querySelector(`#${this.id}`);
     }
 
     private initAutocomplete(): void {
-        this.autocomplete = new autoComplete({
+        this.clearList();
+        // tslint:disable-next-line:no-unused-expression
```
Comment (alfredo-dotcms) [inline projects/dotcms-field-elements/src/components/dot-tags/components/dot-autocomplete/dot-autocomplete.tsx]: why???

File: projects/dotcms-field-elements/src/components/dot-tags/components/dot-chip/dot-chip.e2e.ts [hunk 10]
```diff
@@ -4,105 +4,101 @@ describe('dot-chip', () => {
...
+            spyRemoveEvent = await element.spyOnEvent('remove');
+            await page.waitForChanges();
         });
 
-        it('should trigger remove event', async () => {
-            const spyRemoveEvent = await element.spyOnEvent('remove');
+        describe('remove', () => {
+            it('should trigger', async () => {
+                const button = await getButton();
+                button.triggerEvent('click');
```
Comment (alfredo-dotcms) [inline projects/dotcms-field-elements/src/components/dot-tags/components/dot-chip/dot-chip.e2e.ts]: add `await`

File: projects/dotcms-field-elements/src/components/dot-tags/components/dot-chip/dot-chip.e2e.ts [hunk 11]
```diff
@@ -4,105 +4,101 @@ describe('dot-chip', () => {
...
+                expect(spyRemoveEvent).toHaveReceivedEventDetail('test-tag');
+            });
 
-            expect(spyRemoveEvent).toHaveReceivedEventDetail('test-tag');
+            it('should not trigger', async () => {
+                element.setProperty('disabled', true);
+                await page.waitForChanges();
+
+                const button = await getButton();
+                button.triggerEvent('click');
```
Comment (alfredo-dotcms) [inline projects/dotcms-field-elements/src/components/dot-tags/components/dot-chip/dot-chip.e2e.ts]: add `await`

File: projects/dotcms-field-elements/src/components/dot-tags/dot-tags.e2e.ts [hunk 12]
```diff
@@ -1,348 +1,390 @@
...
 
-        it('should mark any new tag as disabled', async () => {
-            element.setProperty('disabled', true);
+        it('should have touched but pristine', async () => {
+            const autocomplete = await getAutoComplete();
+            autocomplete.triggerEvent('lostFocus', {});
             await page.waitForChanges();
 
-            element.setProperty('value', 'tag-1');
             await page.waitForChanges();
```
Comment (alfredo-dotcms) [inline projects/dotcms-field-elements/src/components/dot-tags/dot-tags.e2e.ts]: remove

File: projects/dotcms-field-elements/src/utils/test/utils.tsx (hunk 1 above)
Comment (fmontes) [reply projects/dotcms-field-elements/src/utils/test/utils.tsx]: Do you expand the `dot-tags.e2e.ts`?

File: projects/dotcms-field-elements/src/components/dot-tags/components/dot-autocomplete/dot-autocomplete.e2e.ts (hunk 6 above)
Comment (fmontes) [reply projects/dotcms-field-elements/src/components/dot-tags/components/dot-autocomplete/dot-autocomplete.e2e.ts]: I want to keep it, we need to find a way to tests blur.

File: projects/dotcms-field-elements/src/components/dot-tags/components/dot-autocomplete/dot-autocomplete.tsx (hunk 9 above)
Comment (fmontes) [reply projects/dotcms-field-elements/src/components/dot-tags/components/dot-autocomplete/dot-autocomplete.tsx]: Is not good to have a `new Something` and not store it anywhere, but in this case the way the autocomplete works we don't need to store the instance anywhere, we just init again when a prop change.

File: projects/dotcms-field-elements/src/components/dot-tags/components/dot-chip/dot-chip.e2e.ts (hunk 10 above)
Comment (fmontes) [reply projects/dotcms-field-elements/src/components/dot-tags/components/dot-chip/dot-chip.e2e.ts]: Screen Shot 2019-06-04 at 3 50 07 PM

triggerEvent doesn't return a promise so we don't need the await, but in this case I change it to a `.click()`.

File: projects/dotcms-field-elements/src/components/dot-tags/components/dot-autocomplete/dot-autocomplete.e2e.ts (hunk 3 above)
Comment (fmontes) [reply projects/dotcms-field-elements/src/components/dot-tags/components/dot-autocomplete/dot-autocomplete.e2e.ts]: Well... I guess, but I was trying to test passing `data` but is not possible, when you pass the function it just doesn't do it, I tried as property and nothing.

**Pull request merged**